[quant][sparsity] Generic convert function

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60728 [quant][sparsity] Generic convert function**
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

Commit:
[quant, sparsity] Generic convert function

[ghstack-poisoned]

Commit:
Update on "[quant, sparsity] Generic convert function"




[ghstack-poisoned]

Commit:
Update on "[quant, sparsity] Generic convert function"




[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"




[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"




[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"




[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

[ghstack-poisoned]

raghuramank100 commented on torch/ao/utils/__init__.py:
```diff
@@ -0,0 +1,109 @@
+    return model
+
+
+def convert(model: nn.Module, mapping: dict = None, config: dict = None) -> nn.Module:
```
For eager mode, the API should only take the config for sparsity, the qconfig is an attribute of the module.  The config is identical to the config arg that is passed to the prepare function of the sparsifier.

```
config = {
        ...         'seq.0.linear1': {
        ...             'zero_block_shape': (1, 4)
        ...         'seq.0.linear2: {
        ...             'zero_block_shape': (1, 4)
        ...         }
        ... }
```

raghuramank100 commented on torch/ao/utils/__init__.py:
```diff
@@ -0,0 +1,109 @@
+        raise AttributeError('FP sparse convert is not yet supported')
+
+    quant_config = config.get('quantized', dict())
+    for mode in config.keys():
```
Lets do this instead.  We can reuse the logic in swap_module for eager mode quantization and keep the configs for 
sparsity and quantization separate
```python


def convert(model: nn.Module, mapping: dict = None, config: dict = None, convert_custom_config_dict:dict=None) -> nn.Module:
if mapping is None:
        mapping = dict()
    if config is None:
        raise AttributeError('Currently, you have to specify the convert config')
    mapping.setdefault('quantized', torch.quantization.get_static_quant_module_class())
    mapping.setdefault(
        'sparse_quantized',
        {
            'static': {nn.Linear, torch.ao.nn.sparse.quantized.Linear},
            'dynamic': {nn.Linear, torch.ao.nn.sparse.quantized.dynamic.Linear},
        })
   if 'sparse' in config:
        raise AttributeError('FP sparse convert is not yet supported')

    if convert_custom_config_dict is None:
        convert_custom_config_dict = {}
    custom_module_class_mapping = convert_custom_config_dict.get("observed_to_quantized_custom_module_class", {})

  model = _joint_convert(model, mapping, config, custom_module_class_mapping)

 def _joint_convert(model, mapping, config, custom_module_class_mapping):
       reassign = {}
       for name, mod in module.named_children():
        # both fused modules and observed custom modules are
        # swapped as one unit
        if not isinstance(mod, _FusedModule) and \
           type(mod) not in custom_module_class_mapping:
            _joint_convert(mod, mapping, config)
            # if mod has qconfig attribute and is present in sparsifier config:
            if hasattr(mod, 'qconfig') 
                 if _module_to_path(mod) in config:
                    mapping = mapping['sparse_quantized']
                 else: 
                    mapping = mapping['quantized']
               # custom module class mapping will not work with sparsity, for now we could set it to None.
                 torch.quantization.swap_module(mod, mapping, custom_module_class_mapping)
                
            
        reassign[name] = swap_module(mod, mapping, custom_module_class_mapping)

    for key, value in reassign.items():
        module._modules[key] = value
 return module

raghuramank100 reviewed (changes requested):
Please see comments

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).

Commit:
Update on "[quant][sparsity] Generic convert function"


This is a wrapper that implements a common convert function for use with quantization and sparsity.
Currently, only the whole model/module conversion is supported.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29465899](https://our.internmc.facebook.com/intern/diff/D29465899)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29465899).
